### PR TITLE
APISIX-38 Allow async plugins

### PR DESCRIPTION
**Pull request by chibenwa**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix (for crazy dudes like I who wants Apisix to run run Redis queries... Blocking the event loop IS a bug. In user code but that's a major source of bugs!)
- [X] New feature provided
- [X] Improve performance (for crazy dudes like I who wants Apisix to run run Redis queries...)

- Related issues https://issues.apache.org/jira/projects/APISIX/issues/APISIX-38

___
### Bugfix

CF https://github.com/apache/apisix-java-plugin-runner/issues/264

The lack of asyn support in this plugin is a major showstopper for doing anything complex.

Turns out I need to call a Redis DBs as part of an OIDC logout workflow cf 

And I need to avoid blocking on the event loop. The current semantic do not allow for this!

CF my first try in https://github.com/linagora/tmail-backend/pull/1412
___
### New feature or improvement

It turns out that callback based APIs are trivial to adapt to asynchronous code.

What we just need is wrap the "write to the client" stuff in the callback. We thus implement the write as a terminal filter. Which gives asynchrony capability for free.

Possible drawbacks:
 - To be noted that the write is likely to happen outside of the event loop, if the user explicitly switched thread. This can have non trivial consequences on the ordering. If that's an issue writes can be forced back to the event loop and be wrapped as an event loop task. Impacts only users that opt-in the async calls though...
 - This means that if client calls the next plugin in an async way we might have an unbounded number of async calls in flight. (Classic async code drawback: your concurrency no longer is bounded by your threadpool...) Impacts only users that opt-in the async calls though... And user that oped in are likely to know what they are doing.
 - The extra plugin array copy. Likely neglictible, and likely movable upon A6Conf initialization but it would be slightly more invasive onto the existing code base...
___
### Next step

 - [x] We will qualify the async plugin strategy in a real world scenario
 - [x] Document how to write an async plugin (example)
 - [x] Document better the trading model (everything runs onto the Netty event loop listening the UNIX socket, DON"T BLOCK)
 - [ ] Happy to write tests if needed. Am I authorized to add something like reactor as a test dependency in order to ease writing tests? Which testing strategy would be best suited to qualify the change?

### CODE REVIEW

**Comment by chibenwa** [comment]
Test failure seems unrelated to my code 

image

**Comment by membphis** [comment]
thx for your contribution, nice job

welcome to review and leave comments if you are good at Java
I am not familiar this language, need community's help

**Comment by chibenwa** [comment]
> welcome to review and leave comments if you are good at Java

I think I have a decent level, so I could get a shot at it.

Question is more about the time though ;)

**Comment by chibenwa** [comment]
Hello all,
First happy new year
Second here are the perf results for My workload: https://github.com/linagora/tmail-backend/pull/1412#issuecomment-2567966186

TL;DR

    1/ Small gains on regular operations (5% mean time AND 7.6 % P99)
    2/ Massive gains in term of ourage (x15 mean time AND x50 P99)

**Comment by hungphan227** [comment]
@chibenwa shouldn't this PR be ready (not draft)?

**Comment by chibenwa** [comment]
Hello Apisix community,

Any update about this?

**Comment by membphis** [comment]
E2E test case failed

pls fix them at first

**Comment by chibenwa** [comment]
`/home/runner/work/_temp/561d5117-b2a0-484e-90f4-8fc8f51c96e2.sh: line 1: docker-compose: command not found`

Please point me to a doc telling how to run them locally because as explained in my email, and above in the PR  the CI environment is broken.

I did run locally `mvn clean install`. Isn't this enough?

**Comment by chibenwa** [comment]
Personal bet: 

The piece of code is so old that we might actually just need to `docker-compose => docker compose`

Though I failed at locating cCI scripts in the ci folder nor in the e2e folder

**Comment by chibenwa** [comment]
The docker worklow needs to be edited:

https://github.com/apache/apisix-java-plugin-runner/actions/runs/12392272043/workflow

I cannot do this within the scope of the PR, admin rights on the repo are needed to do so.

@membphis can you get a shot at this?

**Comment by membphis** [comment]
> The docker worklow needs to be edited:
> 
> https://github.com/apache/apisix-java-plugin-runner/actions/runs/12392272043/workflow
> 
> I cannot do this within the scope of the PR, admin rights on the repo are needed to do so.
> 
> @membphis can you get a shot at this?

you can send me the patch, then I will help you to submit the PR

thx

**Comment by chibenwa** [comment]
I found this file location ^^

And the bet was correct.

I'll write an example and a bit of doc.

**Comment by chibenwa** [comment]
@membphis I think this could be merged...

**Review by moonming — APPROVED** [review]

**Comment by moonming** [comment]
@chibenwa Thank you for your contributions to Apache APISIX 👍 Keep making Java Runner better
